cdisc_rules_engine/operations/get_codelist_attributes.py

```diff
@@ -0,0 +1,172 @@
+        ct_list = ct_cache[(ct_cache[cc].isin(cc_key))]
+        ds_len = self.params.dataframe.shape[0]  # dataset length
+        result = pd.Series([ct_list[cv].values[0] for _ in range(ds_len)])
+        return result
+
+    def _get_ct_from_cache(self, ct_key: str, ct_val: str):
+        """
+        Retrieves the codelist information from the cache based on the given
+        ct_key and ct_val.
+
+        Args:
+            ct_key (str): The key for identifying the codelist.
+            ct_val (str): The value associated with the codelist.
+
+        Returns:
+            pd.DataFrame: A DataFrame containing the codelist information
+            retrieved from the cache.
+        """
+        ct_packages = self.params.ct_packages
+        ct_term_maps = (
+            []
+            if ct_packages is None
+            else [self.cache.get(package) or {} for package in ct_packages]
+        )
+
+        # convert codelist to dataframe
+        ct_result = {ct_key: [], ct_val: []}
+        ct_result = self._add_codelist(ct_key, ct_val, ct_term_maps, ct_result)
+
+        is_contained = set(ct_packages).issubset(set(ct_result[ct_key]))
+        # if all the CT packages exist in Cache, we return the result
+        if is_contained:
+            return pd.DataFrame(ct_result)
+
+        # if not, we need to get them from library
+        config = ConfigService()
+        logger = logging.getLogger()
+        api_key = config.getValue("CDISC_LIBRARY_API_KEY")
+        ct_diff = list(set(ct_packages) - set(set(ct_result[ct_key])))
+
+        cls = CDISCLibraryService(api_key, self.cache)
+        ct_pkgs = cls.get_all_ct_packages()
+        ct_names = [item["href"].split("/")[-1] for item in ct_pkgs]
+
+        for ct in ct_diff:
+            if ct not in ct_names:
+                logger.info(f"Requested package {ct} not in CT library.")
+                continue
+            ct_code = cls.get_codelist_terms_map(ct)
+            ct_result = self._add_codelist(ct_key, ct_val, ct_code, ct_result)
+        return pd.DataFrame(ct_result)
+
+    def _get_ct_from_dataset(self, ct_key: str, ct_val: str):
+        """
+        Retrieves the codelist information from the dataset based on the given
+        ct_key and ct_val.
+
+        Args:
+            ct_key (str): The key for identifying the codelist.
+            ct_val (str): The value associated with the codelist.
+
+        Returns:
+            pd.DataFrame: A DataFrame containing the codelist information
+            retrieved from the dataset.
+        """
+        ct_packages = self.params.ct_packages
+        ct_attr = self.params.ct_attribute  # attribute variable specified in rule
+        c1 = self.params.target  # target variable specified in rule
+        c2 = self.params.ct_version  # controlled term version
+        cc = ct_key  # controlled term variable as key
+        ct_cd = ct_attr
+        if ct_attr == "Term CCODE":
+            ct_cd = "TSVALCD"
```

Comment:
Does this operation need to be TS specific? And can the variables above be named more clearly

Reply:
I think that it is trial summary specific if they specify the attribute is "Term CCODE" otherwise, rule author can use whatever variable as attribute column.